packages/xstate-react/src/useMachineSelector.ts

```diff
@@ -0,0 +1,33 @@
+import {
+  EventObject,
+  StateMachine,
+  Typestate,
+  Interpreter,
+  State
+} from 'xstate';
+import { MaybeLazy } from './types';
+import { useMachine, UseMachineOptions } from './useMachine';
+
+export function useMachineSelector<
```

Comment:
When I've seen this API I thought this would be used for subscribing to an already live machine. I'm wondering if maybe it wouldn't be preferred to bake-in `selector` functionality into `useMachine` instead of providing a dedicated hook for this.

One of the gripes I have with this one is that I don't find the `filter` that is being here in the implementation to be particularly useful - especially that it is enforced here rather than provided as a default. The problem with this approach is that this would most likely **always** lead to rerendering because I doubt that anyone would do a smart caching within `selector`'s implementation and the simple `!==` that is used would just always return `true`.

So given that this is a very small addition to the `useMachine` itself I would rather see this as part of the main hook.

Reply:
> So given that this is a very small addition to the `useMachine` itself I would rather see this as part of the main hook.

The problem is that the return signature is different:

```js
const [state, send] = useMachine(...);
const [partOfState, send] = useMachineSelector(...);
```

Are you saying that you're okay with `state`/`partOfState` being different for the same `useMachine(...)` hook? Not entirely sure how this can be reliably typed.

Reply:
> Are you saying that you're okay with state/partOfState being different for the same useMachine(...) hook? Not entirely sure how this can be reliably typed.

Ye, this sounds OK to me, and seems that it would simplify the API (fewer hooks that do almost the same thing). It indeed raises the complexity of the types though - thanks to conditional types it's doable:
[TS playground](https://www.typescriptlang.org/play?#code/MYGwhgzhAEDKAuZ4FMCyZgAsCWA7ZAPACoDCA9rigB7wA00RAogG7KXTI1sAmMA3tHgBPAA7IAXNAjwATngDm0AL4A+aHyUAoUJBgIkhUhWp0GLNvA5dcvdYNESpshcrUbNmvChkAzDMmgAVQg0DBx8AHkReGwKCGJySk5TJlZ2ZJ5+ezFJaTlcRVV1TWhS6GBjZIB+SSMkmhKykJBkYHga6AAKaQNJfRQEypp6VIsVAEpoAF41AFdcAGtcMgB3XE0tTW5W8BkAn3m22NxoWZD0LDxCRrqTWhvzdOtbAWEcp3zC+5VOxtKAWzCVz6iBQF3ChkSdzMaXgKnu40kAG1+pChilHnCALpbHZgPbQA64I4UU7nIH4Ag3KHJe6lUZPFA2LJvRx5FxKOnQKIxOJWJm2YKhS6RaLHeK3WkwsaaH5-aCAkWOVHgq6DeoY2Hw+VkMVxDo88WaRHQADkzVa8FN-My0AWyCEZB83L1uAg8qqXXm2x8V24NuZLt5bqR5uQLTappxZRj0E9KNBaI1I0xagAPtAAErIeCzGS4IgOAgAOQoxdmIHAACMWgRDXEkQAiC1tRtYlQqaOx0rI7O5-OFsQlssV6u1+sh5vhy1tjtY8by5Go9XQhnYjyaAD0m8EOBgmDA-pEMjINeQ-2gK2w8Ew0AABmdhRDOv96BpxnetzuJzAMoGBC28CSN0iaSGAuBCJMMzQHgPjIDIcDTm0yD+lo26xvGsBISg3A4uhMbAV+3ZlD+AYvFI2EdCBvTQOBkHTGosHwYhEY4coRHEXG0AothKHQBmy6SsM0qUJ2HHEUuiYrlKa5ifhZQLtocSWBeUzQPgKxwImqqUgIPhkGQki4LM-xVsxnJ2KykimsWjAABpENaqidIpFRupYSI9CgpY-li0xkk+VwvgubnSNxXnIH5amPjpyAvm+Sghcp4W8bh-kxRScWvsUMaAcBEVQWonQCBU-wiC0KAgEI1ncNgPhwXslBOQuiUeKFHkRREMhYaxKFRQFsXxTlTTYfliaFWaVZgAAXgoYCmhstH8PKgFURFIIGAQekGUZJlmQhFmvA41m2Q5TkTAxHwKBsC5AA)

This probably has still some room for improvement - it's just a quickly prepared demo.

Reply:
> > Are you saying that you're okay with state/partOfState being different for the same useMachine(...) hook? Not entirely sure how this can be reliably typed.
> 
> Ye, this sounds OK to me, and seems that it would simplify the API (fewer hooks that do almost the same thing). It indeed raises the complexity of the types though - thanks to conditional types it's doable:
> [TS playground](https://www.typescriptlang.org/play?#code/MYGwhgzhAEDKAuZ4FMCyZgAsCWA7ZAPACoDCA9rigB7wA00RAogG7KXTI1sAmMA3tHgBPAA7IAXNAjwATngDm0AL4A+aHyUAoUJBgIkhUhWp0GLNvA5dcvdYNESpshcrUbNmvChkAzDMmgAVQg0DBx8AHkReGwKCGJySk5TJlZ2ZJ5+ezFJaTlcRVV1TWhS6GBjZIB+SSMkmhKykJBkYHga6AAKaQNJfRQEypp6VIsVAEpoAF41AFdcAGtcMgB3XE0tTW5W8BkAn3m22NxoWZD0LDxCRrqTWhvzdOtbAWEcp3zC+5VOxtKAWzCVz6iBQF3ChkSdzMaXgKnu40kAG1+pChilHnCALpbHZgPbQA64I4UU7nIH4Ag3KHJe6lUZPFA2LJvRx5FxKOnQKIxOJWJm2YKhS6RaLHeK3WkwsaaH5-aCAkWOVHgq6DeoY2Hw+VkMVxDo88WaRHQADkzVa8FN-My0AWyCEZB83L1uAg8qqXXm2x8V24NuZLt5bqR5uQLTappxZRj0E9KNBaI1I0xagAPtAAErIeCzGS4IgOAgAOQoxdmIHAACMWgRDXEkQAiC1tRtYlQqaOx0rI7O5-OFsQlssV6u1+sh5vhy1tjtY8by5Go9XQhnYjyaAD0m8EOBgmDA-pEMjINeQ-2gK2w8Ew0AABmdhRDOv96BpxnetzuJzAMoGBC28CSN0iaSGAuBCJMMzQHgPjIDIcDTm0yD+lo26xvGsBISg3A4uhMbAV+3ZlD+AYvFI2EdCBvTQOBkHTGosHwYhEY4coRHEXG0AothKHQBmy6SsM0qUJ2HHEUuiYrlKa5ifhZQLtocSWBeUzQPgKxwImqqUgIPhkGQki4LM-xVsxnJ2KykimsWjAABpENaqidIpFRupYSI9CgpY-li0xkk+VwvgubnSNxXnIH5amPjpyAvm+Sghcp4W8bh-kxRScWvsUMaAcBEVQWonQCBU-wiC0KAgEI1ncNgPhwXslBOQuiUeKFHkRREMhYaxKFRQFsXxTlTTYfliaFWaVZgAAXgoYCmhstH8PKgFURFIIGAQekGUZJlmQhFmvA41m2Q5TkTAxHwKBsC5AA)
> 
> This probably has still some room for improvement - it's just a quickly prepared demo.

Thanks for this; good idea, I'll implement it as a single hook.

Reply:
I've used `=> any` for the `select` and infer the type later but it could also be generic of `UseMachineOptions`. The lack of partial inference is being a real PITA for us as this would be so handy in so many situations... The best would be to write down **how** we expect people to interact with those things and just try to cover those - by using conditional types and providing different overloads we should be able to cover all of the things.

I can do this if we could only write down what are type design goals - like: people should rely on the inference that we provide, etc. Is there anything particular that you'd like to avoid in the type-API that we provide?